### PR TITLE
Initial FLOPs support

### PR DESCRIPTION
**Pull request by pcolange**
This PR adds a new entry into the turnkey_stats file called 'onnx_flops_counter' which holds the total number of FLOPs found in a given ONNX model. It is important to be very specific in how we are defining FLOPs: 

- 1 multiply-accumulate or MAC is a*x + y
- 2 floating-point operations or FLOPs are a*x + y, i.e., FLOPs == 2 * MACs
- There are several repositories that do not make this distinction clearly:
  - [fvcore](https://github.com/facebookresearch/fvcore/blob/e8d19df2ffdaf0a78c8d88d6a2522c36b0cacb07/fvcore/nn/flop_count.py#L55) 
  - [flops-counter](https://github.com/sovrasov/flops-counter.pytorch) (reports MACs)
  - [mmcv](https://github.com/open-mmlab/mmcv/blob/bfdd1f9c4281316f4f654672f0e9dbe1b0624ff0/mmcv/cnn/utils/flops_counter.py#L125) also has a flop counter where they consider 1 flop a mac.
  - [model-summary](https://github.com/ceykmc/pytorch_model_summary/blob/1a8860a45c10921d638f30f87d69436ea3a9cba7/module_madd.py#L4) (in the readme they call these FLOPs)
  - etc...

The concern here is that it appears FLOP is used synonymously with MAC, but most understand the ambiguity and comment their code appropriately.

### CODE REVIEW

File: src/turnkeyml/analyze/util.py [hunk 1]
```diff
@@ -100,6 +100,129 @@ def get_onnx_ops_list(onnx_model) -> Dict:
...
+
+def get_onnx_flops(onnx_model) -> Union[int, None]:
+    """
+    Calculate FLOPs found in the onnx model
+    """
+    try:
+        model = onnx.shape_inference.infer_shapes(
+            onnx.load(onnx_model), strict_mode=True, data_prop=True
+        )
+        # onnx.save(model, "mymodel.onnx")
```
Comment (jeremyfowers) [inline src/turnkeyml/analyze/util.py:131]:
```suggestion
```

File: src/turnkeyml/analyze/util.py [hunk 2]
```diff
@@ -100,6 +100,129 @@ def get_onnx_ops_list(onnx_model) -> Dict:
...
+    elif attr.strings:
+        attr_dict[attr.name] = list(attr.strings)
+    else:
+        attr_dict[attr.name] = "unknown_type"
+    return attr_dict
+
+
+def get_onnx_flops(onnx_model) -> Union[int, None]:
+    """
+    Calculate FLOPs found in the onnx model
```
Comment (jeremyfowers) [inline src/turnkeyml/analyze/util.py:125]: Would you kindly add the caveats from your PR description here?

**Review by jeremyfowers — CHANGES_REQUESTED** [review]
Looking great! I just had a few minor stylistic requests, and we need to handle unsupported FLOP-heavy ops.

Please also verify that your code produces the expected FLOP quantities for a few Transformers, CNNs, and LSTMs and post the result in the PR description.

File: src/turnkeyml/common/filesystem.py [hunk 3]
```diff
@@ -317,6 +317,8 @@ class Keys:
     PARAMETERS = "parameters"
     # Histogram of ONNX operators used in the model
     ONNX_OPS_COUNTER = "onnx_ops_counter"
+    # Number of FLOPs in the model
+    ONNX_FLOPS_COUNTER = "onnx_flops_counter"
```
Comment (ramkrishna2910) [inline src/turnkeyml/common/filesystem.py:321]: Can we add a few lines about this feature to the tool user guide?

File: src/turnkeyml/analyze/util.py [hunk 4]
```diff
@@ -100,6 +100,129 @@ def get_onnx_ops_list(onnx_model) -> Dict:
...
+                if attrs.get("direction") == "bidirectional"
+                else np.int64(1)
+            )
+            seq_length, batch_size, input_dim = inp_dims[0]
+            gate_inp_flops = np.int64(2) * input_dim * hidden_size
+            gate_hid_flops = np.int64(2) * hidden_size * hidden_size
+            num_gates = np.int64(4)
+            unit_flops = num_gates * gate_inp_flops * gate_hid_flops
+            onnx_flops += batch_size * seq_length * direction * unit_flops
+
```
[the remark this answers is not in the document]
Comment (ramkrishna2910) [reply src/turnkeyml/analyze/util.py:222]: +1 on ConvTranspose, ConvInteger and QLinearConv
2 additional candidates to the list above: QuantizeLinear and DequantizeLinear

**Review by ramkrishna2910 — COMMENTED** [review]
Added a couple of comments

File: src/turnkeyml/common/filesystem.py (hunk 3 above)
Comment (jeremyfowers) [reply src/turnkeyml/common/filesystem.py:321]: @ramkrishna2910 is there _any_ documentation of ONNX stats in the tool user guide? You/we might need to start a section documenting the prior stats before Phil has a place to document this one.

File: src/turnkeyml/analyze/util.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (jeremyfowers) [reply src/turnkeyml/analyze/util.py:222]: @ramkrishna2910 I researched QuantizeLinear and DequantizeLinear and I don't think they contain MAC FLOPs. Those ops are meant to convert from fp32 to int8 and back.

[the remark this answers is not in the document]
Comment (pcolange) [reply src/turnkeyml/analyze/util.py:222]: I'll prioritize adding the candidates listed above minus Einsum which needs shape inference support added to ONNX. How do we feel about elementwise ops like Add, Mul, and Div, or others like CumSum, AveragePool, Sigmoid, Softmax etc... I suppose in one case all these ops will affect the FLOPs count, but are mostly negligible compared to larger ops like Conv and Gemm. They might have a bigger impact on graph networks that have a lot of mem ops.

**Comment by pcolange** [comment]
After the latest commit I have verified the following models and will update once I have some transformers added:

image

All of the 3rd party flops were calculated by using [fvcore](https://github.com/facebookresearch/fvcore/tree/main) except for LSTM due to lack of support. I used [THOP](https://github.com/Lyken17/pytorch-OpCounter) for the LSTM models but not the others due to inconsistency with THOP (note the issues on their repo including one from yours truly). Both libraries actually calculate MACs and so I simply multiplied by 2 to get a ballpark of FLOPs. 

The generic LSTM tests changed various parameters like hidden dim, input dim, bidirectional, and num layers. 

Let me know if there are other model candidates to add to the tests. I have a local script that runs the tests and updates the table.

File: src/turnkeyml/analyze/util.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (pcolange) [reply src/turnkeyml/analyze/util.py]: removed comment in latest commits 👍

File: src/turnkeyml/analyze/util.py [hunk 5]
```diff
@@ -100,6 +100,129 @@ def get_onnx_ops_list(onnx_model) -> Dict:
...
+        )
+        # onnx.save(model, "mymodel.onnx")
+    except Exception as e:  # pylint: disable=broad-except
+        printing.log_warning(f"Failed to get ONNX FLOPs from {onnx_model}: {str(e)}")
+        return None
+
+    onnx_flops = np.int64(0)
+    for node in model.graph.node:  # pylint: disable=E1101
+        inp_tensors = {tensor.name: tensor for tensor in model.graph.input}
+        val_tensors = {tensor.name: tensor for tensor in model.graph.value_info}
```
[the remark this answers is not in the document]
Comment (pcolange) [reply src/turnkeyml/analyze/util.py]: Spelled out variables in latest commit 👍

File: src/turnkeyml/analyze/util.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (pcolange) [reply src/turnkeyml/analyze/util.py:243]: I added a condition that skips the FLOPs calculation if there is an unsupported op in the model 👍

File: src/turnkeyml/analyze/util.py [hunk 6]
```diff
@@ -177,13 +300,18 @@ def analyze_onnx(build_name: str, cache_dir: str, stats: fs.Stats):
...
+        onnx_flops_counter = get_onnx_flops(final_onnx_file)
         onnx_model_info = populate_onnx_model_info(final_onnx_file)
         input_dimensions = onnx_input_dimensions(final_onnx_file)
 
         stats.save_model_stat(
             fs.Keys.ONNX_OPS_COUNTER,
             onnx_ops_counter,
         )
+        stats.save_model_stat(
+            fs.Keys.ONNX_FLOPS_COUNTER,
```
[the remark this answers is not in the document]
Comment (pcolange) [reply src/turnkeyml/analyze/util.py]: Made this change in the latest commits 👍

File: src/turnkeyml/analyze/util.py (hunk 2 above)
Comment (pcolange) [reply src/turnkeyml/analyze/util.py]: added in the latest commit 👍

File: src/turnkeyml/analyze/util.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (jeremyfowers) [reply src/turnkeyml/analyze/util.py:299]: > How do we feel about elementwise ops like Add, Mul, and Div, or others like CumSum, AveragePool, Sigmoid, Softmax etc...

My assumption is that they are negligible as well. My suggestion is that if you do a BERT FLOPs count and its 99% accurate without considering those ops, then no need to spend the effort now.

File: src/turnkeyml/analyze/util.py [hunk 7]
```diff
@@ -100,6 +100,205 @@ def get_onnx_ops_list(onnx_model) -> Dict:
...
+        return None
+
+    # If the ONNX model contains one of the following unsupported ops, then this
+    # function will return None since the FLOP total is expected to be incorrect
+    unsupported_ops = [
+        "Einsum",
+        "RNN",
+        "GRU",
+        "DeformConv",
+    ]
```
Comment (jeremyfowers) [inline src/turnkeyml/analyze/util.py:144]: A thought about maintaining this code for the future: if the ONNX spec ever adds a new FLOP-contributing OP then we will need to add it to this list (or add a case for it below). I wonder if there is a good way to remind ourselves to do that? 

Or a good programmatic way to detect ops that emerged in a new opset that this function doesn't support yet? A blunt way would be "if model.opset > MAX_SUPPORTED_OPSET: return None". 

I'm not sure how common it is to get new FLOP-contributing ONNX ops these days. So this comment isn't a blocker for this PR, but something to think about nonetheless.
